### ISSUE TITLE
VMware plugin does not detect VMs with static IPs allocated via dhcpd.conf

### ISSUE THREAD

**Issue by ascotan** [hashicorp/vagrant#8200]
### Vagrant version
Vagrant 1.9.1
vagrant-share (1.1.6, system)
vagrant-vmware-fusion (4.0.15)

### Host operating system
MacOS Sierra 10.12.2
VMware Fusion Version 8.5.3 (4696910

### Guest operating system
centos/7 from hashicorp

### Vagrantfile
```ruby
domain   = '.aaa'

nodes = [
  { :hostname => 'xxx',     :mac => '00:50:56:f0:ca:24'},
  { :hostname => 'yyy',     :mac => '00:50:56:16:c3:77'},
  { :hostname => 'zzz',     :mac => '00:50:56:02:ea:38'},
]

Vagrant.configure("2") do |config|
  config.vm.provider "vmware_fusion"
  nodes.each do |node|
    config.vm.define node[:hostname] do |nodeconfig|
      nodeconfig.vm.box = 'centos/7'
      nodeconfig.vm.hostname = node[:hostname] + domain
      nodeconfig.vm.provider :vmware_fusion do |v|
        v.vmx["memsize"] = '512'
        v.vmx["numvcpus"] = '1'
        v.vmx['ethernet0.addressType'] = 'static'
        v.vmx['ethernet0.address'] = node[:mac]
      end
      nodeconfig.vm.synced_folder ".", "/vagrant", type: "nfs"
      nodeconfig.vm.boot_timeout = 3000
    end
  end
  config.vm.provision "ansible" do |ansible|
    ...
  end
end
```

### Debug output
vagrant up xxx
Bringing machine 'xxx' up with 'vmware_fusion' provider...
==> apis: Cloning VMware VM: 'centos/7'. This can take some time...
==> apis: Checking if box 'centos/7' is up to date...
==> apis: Verifying vmnet devices are healthy...
==> apis: Preparing network adapters...
==> apis: Fixed port collision for 22 => 2222. Now on port 2210.
==> apis: Starting the VMware VM...
==> apis: Waiting for machine to boot. This may take a few minutes...
Timed out while waiting for the machine to boot. This means that
Vagrant was unable to communicate with the guest machine within
the configured ("config.vm.boot_timeout" value) time period.

If you look above, you should be able to see the error(s) that
Vagrant had when attempting to connect to the machine. These errors
are usually good hints as to what may be wrong.

If you're using a custom box, make sure that networking is properly
working and you're able to connect to the machine. It is a common
problem that networking isn't setup properly in these boxes.
Verify that authentication configurations are also setup properly,
as well.

If the box appears to be booting properly, you may want to increase
the timeout ("config.vm.boot_timeout") value.

### Expected behavior
Vagrant detects the IP of eth0 and logs in for provisioning

### Actual behavior
Vagrant never detects the IP even though VMware has assigned an IP to eth0

### Steps to reproduce
1. Create a Vagrantfile like the above
2. Edit vmnet8:
```
sudo vi /Library/Preferences/VMware\ Fusion/vmnet8/dhcpd.conf
```
and add the following below the 'do not edit' section:
```
host xxx.aaa {
    hardware ethernet 00:50:56:f0:ca:24;
    fixed-address 192.168.247.100;
}
host yyy.aaa {
    hardware ethernet 00:50:56:16:c3:77;
    fixed-address 192.168.247.101;
}
host zzz.aaa {
    hardware ethernet 00:50:56:02:ea:38;
    fixed-address 192.168.247.102;
}
```
These IPs will fall into vmnet8's static allocation block for the 247 subnet. (You can check this by looking at the dhcpd.conf file and seeing that the subnet's dhcp range is 128+
3. Restart VMware networking:
```
sudo /Applications/VMware\ Fusion.app/Contents/Library/vmnet-cli --stop
sudo /Applications/VMware\ Fusion.app/Contents/Library/vmnet-cli --configure
sudo /Applications/VMware\ Fusion.app/Contents/Library/vmnet-cli --start
```
2. Run 'vagrant up'
3. The machine never boots looking for a DHCP reservation
4. Open VMWare Fusion and log into the VM via the GUI with the user/pass vagrant/vagrant.
5. Verify that the network is up and that vmnet8 has given the VM an IP on eth0:
```
nmcli d show eth0
```

As a workaround for this I have to add the static IP to the dhcp lease database temporarily on every VM creation to get vagrant to find the IP of the VM. You can do this by editing vmnet8's dhcp lease database:
```
sudo vi /var/db/vmware/vmnet-dhcpd-vmnet8.leases
```
and putting in entries for DHCP leases outside of it's dhcp reservation range that match the static IPs:
```
lease 192.168.247.100 {
        starts 0 2017/01/01 00:00:00;
        ends never;
        hardware ethernet 00:50:56:f0:ca:24;
        client-hostname "xxx.aaa";
}
lease 192.168.247.101 {
        starts 0 2017/01/01 00:00:00;
        ends never;
        hardware ethernet 00:50:56:16:c3:77;
        client-hostname "yyy.aaa";
}
lease 192.168.247.102 {
        starts 0 2017/01/01 00:00:00;
        ends never;
        hardware ethernet 00:50:56:02:ea:38;
        client-hostname "zzz.aaa";
}
```

The will allow the VM to be created, but every machine must be booted separately because VMware realizes that these leases are bogus and deletes them during VM provisioning.
Strangely, however, once the VM is up, vagrant can determine what the IP of the VM was and 'vagrant ssh' works even though the fake leases are gone.

As a sidenote adding a static IP via something like:
config.vm.network "private_network", ip: "192.168.50.4" 
appears to also not work and adds another network interface called eth1 leaving eth0 broken....

**Comment by kikitux** [comment]
@ascotan 

Can you comment out or remove this:

```
v.vmx['ethernet0.addressType'] = 'static'
```

Even if the ip is fixed for the given mac, its still using dhcp

**Comment by ascotan** [comment]
I commented out the 'static' addressType and the VM boots, but the VM is give a DHCP lease with an IP above 128 in the DHCP range. The MAC address is not given to eth0 and therefore the wrong IP is allocated.

```
nmcli d show eth0
GENERAL.DEVICE:                         eth0
GENERAL.TYPE:                           ethernet
GENERAL.HWADDR:                         00:0C:29:49:5A:DB
GENERAL.MTU:                            1500
GENERAL.STATE:                          100 (connected)
GENERAL.CONNECTION:                     System eth0
GENERAL.CON-PATH:                       /org/freedesktop/NetworkManager/ActiveConnection/0
WIRED-PROPERTIES.CARRIER:               on
IP4.ADDRESS[1]:                         192.168.247.128/24
IP4.GATEWAY:                            192.168.247.2
IP4.DNS[1]:                             192.168.247.2
IP4.DOMAIN[1]:                          localdomain
IP6.ADDRESS[1]:                         fe80::20c:29ff:fe49:5adb/64
IP6.GATEWAY:
```

**Comment by ntrippar** [comment]
im having the same issue. with every linux distribution hangs trying to connect to vm . i checked and its trying to get the IP from the vmware file and the ip isnt there. one way to solve the issue is open the vmware fusion gui. but isnt usable. any news about this?

**Comment by jstortz** [comment]
Same exact issue. Report by ascotan was fabulously detailed! I am experiencing the same thing.

**Comment by chrisroberts** [comment]
Mappings won't persist due to the dhcpd.conf file being rewritten on networking re-configurations causing the custom mappings to be erased. In recent versions of the vmware plugin support has been added for the `base_address` configuration option:

https://www.vagrantup.com/docs/vagrantfile/machine_settings.html#config-vm-base_address

This allows you to define an address for the guest and the dhcp mapping will be created correctly for the guest providing the guest with the defined address. Cheers!

**Comment by mweisel** [comment]
Is this setting valid for the latest **vagrant-vmware-desktop** plugin (2.0.3)?

```
config.vm.define "NX1" do |nx|
    #nx.ssh.host = "192.168.200.11"
    #nx.vm.base_mac = "00:50:56:31:20:90"
    nx.vm.base_address = "192.168.200.11"
    nx.ssh.shell = "run bash"
    nx.vm.box = "cisco-nexus-9kv-703-i7-5a"
    # Set timeout to 6 mins
    nx.vm.boot_timeout = 360
...
```

```
$ vagrant validate
There are errors in the configuration of this machine. Please fix
the following errors and try again:

vm:
* The following settings shouldn't exist: base_address
```

If not, what is the current solution? Thanks.

FYI, the mappings are persistent with VMware Fusion 11. The Cisco Nexus virtual appliance uses the reservation correctly, but Vagrant hangs _waiting_ for the IP address.

```
Bringing machine 'NX1' up with 'vmware_desktop' provider...
Bringing machine 'NX2' up with 'vmware_desktop' provider...
==> NX1: Cloning VMware VM: 'cisco-nexus-9kv-703-i7-5a'. This can take some time...
==> NX1: Verifying vmnet devices are healthy...
==> NX1: Preparing network adapters...
==> NX1: Starting the VMware VM...
==> NX1: Waiting for the VM to receive an address...
```

```
nx.vm.provider :vmware_desktop do |v|
... 
 # Set to avoid duplicate UID DHCP conflict
  v.vmx["ethernet0.addressType"] = "static"
  v.vmx["ethernet0.address"] = "00:50:56:31:20:90"
...
end
```

```
$ tail -5 /Library/Preferences/VMware\ Fusion/vmnet8/dhcpd.conf
####### VMNET DHCP Configuration. End of "DO NOT MODIFY SECTION" #######
host NX1 {
    hardware ethernet 00:50:56:31:20:90;
    fixed-address 192.168.200.11;
}
```

`$ ssh -i $HOME/.vagrant.d/insecure_private_key vagrant@192.168.200.11`

```
Cisco NX-OS Software
Copyright (c) 2002-2018, Cisco Systems, Inc. All rights reserved.
Nexus 9000v software ("Nexus 9000v Software") and related documentation,
files or other reference materials ("Documentation") are
the proprietary property and confidential information of Cisco
Systems, Inc. ("Cisco") and are protected, without limitation,
pursuant to United States and International copyright and trademark
laws in the applicable jurisdiction which provide civil and criminal
penalties for copying or distribution without Cisco's authorization.

Any use or disclosure, in whole or in part, of the Nexus 9000v Software
or Documentation to any third party for any purposes is expressly
prohibited except as otherwise authorized by Cisco in writing.
The copyrights to certain works contained herein are owned by other
third parties and are used and distributed under license. Some parts
of this software may be covered under the GNU Public License or the
GNU Lesser General Public License. A copy of each such license is
available at
http://www.gnu.org/licenses/gpl.html and
http://www.gnu.org/licenses/lgpl.html
***************************************************************************
*  Nexus 9000v is strictly limited to use for evaluation, demonstration   *
*  and NX-OS education. Any use or disclosure, in whole or in part of     *
*  the Nexus 9000v Software or Documentation to any third party for any   *
*  purposes is expressly prohibited except as otherwise authorized by     *
*  Cisco in writing.                                                      *
***************************************************************************
switch# sh ip int br vrf management

IP Interface Status for VRF "management"(2)
Interface            IP Address      Interface Status
mgmt0                192.168.200.11  protocol-up/link-up/admin-up
```
